### ISSUE TITLE
Locking dependencies is slow

### ISSUE THREAD

**Issue by cab404** [svanderburg/node2nix#264]
Running node2nix with `package-lock.json` takes too much time. It looks like fetching is not parallelized whatsoever.

Are there any opportunities to speed things up?

**Comment by svanderburg** [comment]
Yes it was a decision made in 2014, and node2nix is still influenced by that decision.

The good news is that I already a local PoC implementation that I can use to deploy NPM projects, so most of the hard work is already done. 

So the idea of my re-architecting effort is to make its design more modular. There will be separated pieces each taking a specific responsibility: fetching dependency metadata, performing NPM's dependency resolution algorithm, installing NPM dependencies in a build environment etc.

There will also be a library/package that specifically handles project deployments that I intend to publish in the next few weeks. Once this is available, it can be tested.
